src/renderers/dom/shared/ReactDOMComponent.js

```diff
@@ -632,12 +633,21 @@ ReactDOMComponent.Mixin = {
           this._currentElement.type
         );
       }
+      var isCustomComponentTag = isCustomComponent(this._tag, props);
+      if (__DEV__ && isCustomComponentTag && !didWarnShadyDom) {
+        warning(
+          !el.shadyRoot,
+          'A component is using shady dom. Using shady dom with React can ' +
+          'cause things to break subtly.'
+        );
+        didWarnShadyDom = true;
```

Comment:
Ah good catch! I went ahead and added the check for `&& el.shadyRoot` and replaced the `warning` argument with `didWarnShadyDom`